### PR TITLE
Common build provider to hide specific build system

### PR DESCRIPTION
**Pull request by vhatsura**
Base implementation for #259

### CODE REVIEW

**Comment by gep13** [comment]
@pascalberger can I get your thoughts on this?  Thanks

File: Cake.Recipe/Content/buildProvider.cake [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+public interface IBuildProvider
+{
+    IRepositoryInfo Repository { get; }
+
+    IPullRequestInfo PullRequest { get; }
+
+    IBuildInfo Build { get; }
+}
+
+public class AppVeyorTagInfo : ITagInfo
```
[the remark this answers is not in the document]
Comment (vhatsura) [reply Cake.Recipe/Content/buildProvider.cake]: Fixed

File: Cake.Recipe/Content/buildProvider.cake [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    public IRepositoryInfo Repository { get; }
+
+    public IPullRequestInfo PullRequest { get; }
+
+    public IBuildInfo Build { get; }
+}
+
+public static IBuildProvider GetBuildProvider(ICakeContext context, BuildSystem buildSystem)
+{
+    // always fallback to AppVeyor
```
[the remark this answers is not in the document]
Comment (pascalberger) [reply Cake.Recipe/Content/buildProvider.cake]: Agreed 👍

**Comment by pascalberger** [comment]
@vhatsura Can you rebase on current develop or check the `Allow edits from maintainers.` checkbox on the right side of this PR?

**Comment by vhatsura** [comment]
@pascalberger, `Allow edits from maintainers` was checked and didn't changed. Will rebase on current develop

**Comment by pascalberger** [comment]
@vhatsura Are you sure [Allow edits from maintainers](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) is enabled? I get permission denied when I try to write to the branch. Or maybe you have some branch policy enabled?

It would be nice to have it rebased before merging, to get rid of the merge commits (you updated the branch with an additional merge commit). While I could squash together the commits, to get rid of them, GitHub will lost track that these changes are made by you and attribute them to me (if this is OK for you I can also do it this way).

**Comment by vhatsura** [comment]
@pascalberger, hmm, yes I'm sure that checkbox is checked. Also I checked branch and repo policies. All looks fine.

By the way, I did rebase onto upstream.

**Comment by pascalberger** [comment]
@vhatsura Sorry, there was some config issue on my side. I've now cleanup the branch (you were creating merge commits from upstream `develop` into your develop branch, not rebasing on upstream develop). 

The changes have now been merged, thanks for your contribution 👍

For the future GitHub workflow is easier if you make changes on a separate branch in your fork, as now the `develop` branch in your fork and the one in the upstream repository has diverged. See [GitHub Flow](https://guides.github.com/introduction/flow/).

**Pull request merged**